src/js/components/Form/Form.js

```diff
@@ -0,0 +1,129 @@
+import React, { Component } from 'react';
+import { defaultProps } from '../../default-props';
+import { deepMerge } from '../../utils';
+import { FormContext } from './FormContext';
+
+const defaultMessages = {
+  invalid: 'invalid',
+  required: 'required',
+};
+
+class Form extends Component {
+  static defaultProps = {
+    messages: defaultMessages,
+    value: {},
+  };
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    const { value, errors, messages } = nextProps;
+    const {
+      value: stateValue,
+      errors: stateErrors,
+      priorValue,
+      priorErrors,
+      priorMessages,
+    } = prevState;
+    if (
+      !priorValue ||
+      value !== priorValue ||
+      errors !== priorErrors ||
+      messages !== priorMessages
+    ) {
+      return {
+        value: value !== priorValue ? value : stateValue,
+        priorValue: value,
+        errors: (errors !== priorErrors ? errors : stateErrors) || {},
+        priorErrors: errors,
+        messages: deepMerge(defaultMessages, messages),
```

Comment:
why do we need a deepMerge here ?

Reply:
I was allowing the caller to pass sparse `messages`.

Reply:
i hear you, but does it need to be "deep"?

Reply:
doh, of course not. Fixed.